ci: AppVeyor fixes for Visual Studio 2019 16.8.1 image

sipsorcery opened the pull request:
The motivation for this PR is twofold:

  1. Update the Qt binaries used by the appveyor CI job after a recent update to Visual Studio 2019 used in the Appveyor build image resulted in ABI incompatibilities,

  2. Remove optimisations and debug information from the Bitcoin Core `Release` msvc build to reduce the chance of future ABI incompatibility issues for future Visual Studio updates.

The changes made in this PR are:

 - Changed appveyor config file hash to use a new version of Qt pre-compiled binaries built for Visual Studio 2019 v16.8.1.
 - Adjusted msvc compiler and linker settings to remove optimisations and debug information generation to help avoid future ABI issues on Visual Studio updates.
 - Tidied up debug and release configuration blocks in common project file to avoid duplication.
 - Updated appveyor config to use latest Visual Studio 2019 image.
 - Bumped vcpkg version to tag `2020.11-1` for binary caching feature*.

See #20392 for related discussion.

*Binary caching is a new [vcpkg feature](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/binarycaching.md) that allows dependency caching. This PR is not using the feature but by updating the vcpkg version it means it can be optionally used by other contributors in their own Appveyor configs. By caching the vcpkg dependencies using this feature my build times are reduced by approx 10 minutes.

adamjonas commented on build_msvc/common.init.vcxproj:
```diff
@@ -45,26 +45,27 @@
-    <LinkIncremental>false</LinkIncremental>
-    <WholeProgramOptimization>true</WholeProgramOptimization>
-    <UseDebugLibraries>false</UseDebugLibraries>
+  
```
Heads up that the linter is complaining about the two spaces at the start of this line.

sipsorcery replied:
Thanks.

sipsorcery committed:
Adjusted msvc compiler and linker settings to remove optimisations that are causing sporadic ABI issues on Visual Studio updates.

Tidied up debug and release configuration blocks in common project file to avoid duplication.

Updated appveyor config to use latest Visual Studio 2019 image.

Changed appveyor config file hash to use a new version of Qt pre-compiled binaries built for Visual Studio 2019 v16.8.1.

Bumped vcpkg version to tag '2020.11-1' for binary caching feature.

See #20392 for related discussion.

sipsorcery commented:
@MarcoFalke as a follow on to your comment [here](https://github.com/bitcoin/bitcoin/pull/19960#issuecomment-718087777)  this PR should reduce the appveyor build times by approximately 20 minutes. From around 65 to around 45.

The reduction is down to building the Qt libraries without msvc compiler optimisations. As mentioned above, removing the optimisations was to help with ABI compatibility. The build time reduction was a welcome side effect.

hebasto reviewed (approved):
ACK 8b99e609e7da5dd3601e9214d8f869e96108fffe, [tested](https://ci.appveyor.com/project/hebasto/bitcoin/builds/36629021) in my personal repo; did not check changes in the `build_msvc/` directory.

AppVeyor finished build in 46 min 31 sec.

sipsorcery commented:
@hebasto thanks for checking. If you have time could you do me a favour (it's a quick set up) and check the vcpkg binary caching feature on your appveyor account?

If so the steps are:
 - On your appveyor project for bitcoin add two environment variables (this is on the appveyor settings not in the yml file):
  VCPKG_FEATURE_FLAGS manifests,binarycaching
  VCPKG_BINARY_SOURCES nuget,https://aaronrc.pkgs.visualstudio.com/bitcoin/_packaging/vcpkg-deps/nuget/v3/index.json,read
 - Re-run the same commit you ran above.

![appveyor_env](https://user-images.githubusercontent.com/197660/100935274-cd37ad80-34e7-11eb-8d0c-afddeb4e886d.png)

The test is to see whether your build job pulls the pre-built vcpkg dependencies from my Azure devops account. The reason it may not is the way vcpkg generates the hash to determine whether the pre-built dependency is a match or not.

If the pre-built dependencies are used your build time should be closer to 30 minutes.

MarcoFalke commented on .appveyor.yml:
```diff
@@ -1,28 +1,29 @@
-      git pull origin master > $null
-      git -c advice.detachedHead=false checkout $env:VCPKG_COMMIT_ID
+      git -c advice.detachedHead=false checkout $env:VCPKG_TAG
+      git pull origin $env:VCPKG_TAG
```
Why does it first checkout the tag and then pull it? Shouldn't it be the other way round?

sipsorcery committed:
Removed redundant git pull from appveyor config.

MarcoFalke commented:
Concept ACK 2c69381f3de5091e103cb8bef299aba321503e7c

Haven't reviewed the code, nor have I tested if building locally on Windows still works

hebasto commented:
@sipsorcery 
> @hebasto thanks for checking. If you have time could you do me a favour (it's a quick set up) and check the vcpkg binary caching feature on your appveyor account?

Configured and checked: re-build has finished "in 30 min".

sipsorcery commented:
> Configured and checked: re-build has finished "in 30 min".

Perfect, thanks!

laanwj commented:
Concept ACK 2c69381f3de5091e103cb8bef299aba321503e7c
Checked that this only touches the MSVC files.

sipsorcery commented:
@MarcoFalke if you think it's a good idea you can add the environment setting below to the Bitcoin Core appveyor job to further reduce the build time by between 10 and 15 minutes. As mentioned above this setting is better put in the Appveyor build job settings rather than in the `yml` file.

`VCPKG_BINARY_SOURCES` `nuget,https://aaronrc.pkgs.visualstudio.com/bitcoin/_packaging/vcpkg-deps/nuget/v3/index.json,read`

Unlike previous attempts at caching vcpkg dependencies this one should be more reliable:

 - If a particular package doesn't have a matching nuget package it will be built from source,
 - If a new dependency is added to the `vcpkg.json` file it will be built from source,
 - If the Appveyor image gets updated the version hash of the vcpkg dependency changes and will cause it to be built from source,
 - The stored nuget packages are treated as readonly with the above environment variable. It requires a manual script to be run to add new packages.

I've created a custom script to run as a standalone appveyor build job which builds the dependencies in `vcpkg.json` and stores the results as nuget packages in my Azure site. It would be trivial to switch to a different nuget storage location or even standard file storage if required. It's probably worth verifying that this approach does prove more reliable than previous attempts before expending too much effort.
